### PR TITLE
Various understanding tweaks

### PR DESCRIPTION
**Pull request by mikenomitch**
This PR makes several tweaks to the compat matrix to improve understanding:

- Adds information on which Cloudflare column equates to which node(js)_compat version
- Replaces raw API numbers with percentages for missing, mismatch, and stub. (because the raw number only means something in the context of the total IMHO)
- Adds information in notes to help the user understand columns and definitions
- Adds a subtitle explaining the "/"ed numbers in each cell
- Adds disclaimer about data being generated via script
- Rename Supported -> Present, Unsupported -> Missing, Stub -> Mock

Note: I am a tad worried about all the percentage calculations making rendering slow on expand and/or initial page load. I can revert this change if that's an issue - or alternatively update the data to include these percentages so we have them pre-calculated.

Screenshot 2024-08-29 at 10 00 53 AM

### CODE REVIEW

File: report/src/App.tsx [hunk 1]
```diff
@@ -289,15 +317,29 @@ const App = () => {
           <h3 className="font-semibold text-lg mb-1">Notes</h3>
           <ul className="list-disc list-inside">
             <li className="mb-1">
-              All percentages in the table represent API presence whether the shape of the API 
-              matches against the baseline. They are not a calculation of implementation compliance. 
+              All percentages in the table represent API presence whether the
```
Comment (vicb) [inline report/src/App.tsx:320]:
```suggestion
              All percentages in the table represent whether the
```

File: report/src/utils.ts [hunk 2]
```diff
@@ -44,7 +44,7 @@ export const pct = (part: number, total: number) => {
   return (part / total) * 100;
 };
 
-export const formatPct = (percentage: number) => {
+export const formatPct = (percentage: number, hideSymbol: boolean = false) => {
```
Comment (vicb) [inline report/src/utils.ts:47]:
```suggestion
export const formatPct = (percentage: number, hideSymbol = false) => {
```

**Comment by vicb** [comment]
> I am a tad worried about all the percentage calculations making rendering slow

One thing we could do is to compute the percentages when we generate the json & csv files so that the rendering is not slowed down.

File: report/src/App.tsx [hunk 3]
```diff
@@ -14,6 +14,12 @@ import { z } from "zod";
 const RED_THRESHOLD = 20;
 const GREEN_THRESHOLD = 80;
 
+const versionSubtitles = {
+  workerd: "node_compat = true",
```
Comment (IgorMinar) [inline report/src/App.tsx:18]:
```suggestion
  workerd: "nodejs_compat",
```

File: report/src/App.tsx [hunk 4]
```diff
@@ -14,6 +14,12 @@ import { z } from "zod";
...
+  wranglerV3: "nodejs_compat",
```
Comment (IgorMinar) [inline report/src/App.tsx:19]:
```suggestion
  wranglerV3: "(legacy) node_compat = true",
```

File: report/src/App.tsx [hunk 5]
```diff
@@ -116,14 +122,22 @@ const App = () => {
               .split("/")
               .map((i) => parseInt(i as string));
 
-            const total = supported + mismatch + stub;
-            const percentage = pct(total, baselineSupport as number);
+            const totalImplemeneted = supported + mismatch + stub;
```
Comment (IgorMinar) [inline report/src/App.tsx:125]: I suggest we move away from "implemented" or "supported" and instead use "present" and "missing" as the terminology to describe if an API exists in workers or not.

File: report/src/App.tsx [hunk 6]
```diff
@@ -116,14 +122,22 @@ const App = () => {
...
+            const total = totalImplemeneted + unsupported;
+            const percentageImplemented = pct(
+              totalImplemeneted,
+              baselineSupport as number
+            );
+
+            const unsupportedPct = pct(unsupported, total);
```
Comment (IgorMinar) [inline report/src/App.tsx:132]: "unsupported" -> "missing"

File: report/src/utils.ts [hunk 7]
```diff
@@ -44,7 +44,7 @@ export const pct = (part: number, total: number) => {
   return (part / total) * 100;
 };
 
-export const formatPct = (percentage: number) => {
+export const formatPct = (percentage: number, hideSymbol = false) => {
```
Comment (IgorMinar) [inline report/src/utils.ts:47]: I don't quite understand why do you want to hide the % symbol. Could you please help me here?

I find the part of the ui without "%" captured in the screenshot puzzling. I suggest you remove this part and use % everywhere.

File: report/src/App.tsx (hunk 3 above)
Comment (IgorMinar) [reply report/src/App.tsx:18]: Good call on adding the activation flags here.

Except, you had the flags a bit wrong wrong, so I suggested fixes to match what each column really represents.

File: report/src/App.tsx [hunk 8]
```diff
@@ -289,15 +317,29 @@ const App = () => {
...
             <li className="mb-1">
-              All percentages in the table represent API presence whether the shape of the API 
-              matches against the baseline. They are not a calculation of implementation compliance. 
+              All percentages in the table represent whether the
+              shape of the API matches against the baseline. They are not a
+              calculation of implementation compliance.
+            </li>
+            <li className="mb-1">
+              The percentages represent the API surface area that is supported{" "}
+              <span className="font-semibold">or</span> stubbed{" "}
```
Comment (IgorMinar) [inline report/src/App.tsx:326]: In the meeting today, we agreed to rename "stub" with "mock" as that's the terminology that unenv uses, so we might as well follow that precedence. Could you please update here and elsewhere in UI and code where we say "stub"? thank you

File: report/src/App.tsx [hunk 9]
```diff
@@ -289,15 +317,29 @@ const App = () => {
...
+              <span className="font-semibold">or</span> mismatched.
+            </li>
+            <li className="mb-1">
+              The <span className="font-semibold">baseline</span> column
+              represents a union of Node.js v18, v20, and v22 API surfaces that
+              we use as node API compatibility target.
```
Comment (IgorMinar) [inline report/src/App.tsx:332]:
```suggestion
              we use as the ideal Node.js API compatibility target.
```

**Review by IgorMinar — COMMENTED** [review]
Thanks for the improvements!

I suggest that we update some of the flags and terms, but don't try to make major UX changes (like dropping % from some but not all of the columns) as none of us are skilled to do that well. If we really care about this looking good, we should ask Nena for help. I'm sure that someone on her team could spend just a bit of time and make the UI meaningfully better.

File: report/src/utils.ts (hunk 7 above)
Comment (mikenomitch) [reply report/src/utils.ts:47]: I initially had it and then it felt a bit unnecessary and took up space.

I wavered a bit on whether to drop it or not, so it's good feedback if it feels off to you. I can revert this bit.

File: report/src/App.tsx (hunk 5 above)
Comment (mikenomitch) [reply report/src/App.tsx:125]: Yeah, I saw the wording changes you all suggested in the doc after pushing this - can change

File: report/src/App.tsx (hunk 3 above)
Comment (mikenomitch) [reply report/src/App.tsx:18]: Ah whoops! All the more reason to add them for others :)

File: report/src/App.tsx (hunk 8 above)
Comment (mikenomitch) [reply report/src/App.tsx:326]: +1 on this

**Comment by mikenomitch** [comment]
Thanks for the review @IgorMinar - will hop on these changes tomorrow

**Comment by mikenomitch** [comment]
@IgorMinar updated - check new screenshot in top level description

File: report/src/App.tsx [hunk 10]
```diff
@@ -289,32 +307,53 @@ const App = () => {
           <h3 className="font-semibold text-lg mb-1">Notes</h3>
           <ul className="list-disc list-inside">
             <li className="mb-1">
-              All percentages in the table represent API presence whether the shape of the API 
-              matches against the baseline. They are not a calculation of implementation compliance. 
+              All percentages in the table represent API presence whether the
+              shape of the API matches against the baseline. They are not a
+              calculation of implementation compliance.
```
Comment (vicb) [inline report/src/App.tsx:312]:
```suggestion
              All percentages in the table represent whether the
              shape of the API matches against the baseline.
              <span className="font-bold">They are not a
              calculation of implementation compliance.</span>
```

File: report/src/App.tsx [hunk 11]
```diff
@@ -221,26 +236,29 @@ const App = () => {
           </div>
         </TableCell>
         {targetTotals.map((targetTotal) => {
-          const [supported, mismatch, stub, unsupported] = (
-            targetTotal as string
-          )
+          const [matching, mismatch, mock, missing] = (targetTotal as string)
             .split("/")
             .map((i) => parseInt(i as string));
```
Comment (vicb) [inline report/src/App.tsx:241]:
```suggestion
            .map((i) => parseInt(i as string, 10));
```

File: report/src/App.tsx [hunk 12]
```diff
@@ -221,26 +236,29 @@ const App = () => {
...
-          const percentage = pct(total, baselineCount as number);
+          const totalPresent = matching + mismatch + mock;
+          const total = totalPresent + missing;
+          const presentPct = pct(totalPresent, baselineCount as number);
+          const missingPct = pct(missing, total);
+          const mismatchPct = pct(mismatch, total);
+          const mockPct = pct(mock, total);
 
-          const tooltip = `Unsupported: ${unsupported}\nMismatch: ${mismatch}\nStubbed: ${stub}\nSupported: ${supported}`;
+          const tooltip = `Missing: ${missing}\nMismatch: ${mismatch}\nMocked: ${mock}\nMatching: ${matching}`;
```
Comment (vicb) [inline report/src/App.tsx:250]:
```suggestion
          const tooltip = `Matching: ${matching}\nMissing: ${missing}\nMismatch: ${mismatch}\nMocked: ${mock}`;
```

File: report/src/App.tsx [hunk 13]
```diff
@@ -221,26 +236,29 @@ const App = () => {
...
                 <span className="text-sm font-semibold">
-                  {formatPct(percentage)}
+                  {formatPct(presentPct)}
                 </span>
                 <div className="text-xs">
                   <span>
-                    {unsupported}/{mismatch}/{stub}
+                    {formatPct(missingPct)}/{formatPct(mismatchPct)}/
+                    {formatPct(mockPct)}
                   </span>
```
Comment (vicb) [inline report/src/App.tsx:262]: I think we should remove this - the amount of info in the table is overwhelming.

I'd rather use a proper tooltip to show the info on hover - using `title` has the drawback of introducing a small but anoying delay before the tooltip is displayed.

Quick Google search for better tooltips:
- https://github.com/cosbgn/tailwindcss-tooltips
- https://flowbite.com/docs/components/tooltips/
- https://www.material-tailwind.com/docs/html/tooltip
- https://pagedone.io/docs/tooltips

**Review by vicb — CHANGES_REQUESTED** [review]
I added a few inline comments.

If we have "Missing" should we have "Present" instead of "Matching"?
IMO "Present" better reflect that the behavior does not necesarily matches Node's behavior.

File: report/src/App.tsx (hunk 13 above)
Comment (mikenomitch) [reply report/src/App.tsx:262]: I agree it's pretty overwhelming, but also I didn't realize there were tooltips until actually seeing them in the code. So discoverability takes a big hit with just that. And IMO these percentages are pretty important for full understanding.

Mind if we merge this as an iterative improvement and then try to figure out a better solution later?

Comment (vicb) [reply report/src/App.tsx:262]: > I agree it's pretty overwhelming, but also I didn't realize there were tooltips until actually seeing them in the code.

Same here :)
Which is why a proprer tooltip would help.

But yes, we can merge and figure out later

**Review by vicb — APPROVED** [review]

**Pull request merged**